### PR TITLE
Use single stream for apply optimizer in backward

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89963
* #89194

Per title

Differential Revision: [D41622106](https://our.internmc.facebook.com/intern/diff/D41622106/)

### CODE REVIEW

File: torch/distributed/algorithms/ddp_comm_hooks/optimizer_overlap_hooks.py [hunk 1]
```diff
@@ -40,23 +49,55 @@ def _apply_optim_in_backward_hook(
...
+            optim_stream_state.wait_for_optim_stream_enqueued = False
+
+        if not optim_stream_state.wait_for_optim_stream_enqueued:
+            Variable._execution_engine.queue_callback(
+                wait_for_optim_stream_callback
+            )
+            # mark that the callback is enqueued
+            optim_stream_state.wait_for_optim_stream_enqueued = True
+
+        return fut
```
Comment (zhaojuanmao) [inline torch/distributed/algorithms/ddp_comm_hooks/optimizer_overlap_hooks.py:91]: we do not need to queue call back, instead we can simply call " torch.cuda.current_stream().wait_stream(
                optim_stream_state.optim_stream)" in the final_backward call back?

**Review by zhaojuanmao — APPROVED** [review]
left one comment

File: torch/distributed/algorithms/ddp_comm_hooks/optimizer_overlap_hooks.py (hunk 1 above)
Comment (rohan-varma) [reply torch/distributed/algorithms/ddp_comm_hooks/optimizer_overlap_hooks.py:91]: Is this referring to calling this in `finalize_backward` in C++? I agree we can do this, I am just not sure if we can pass a cuda stream from python to C++, but it should be possible.

**Comment by rohan-varma** [comment]
close in favor of https://github.com/pytorch/pytorch/pull/89194

**Pull request closed without merging**